app/services/checkout/form_data_adapter.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# Adapts checkout form data (params) so that the order can be directly saved to the database
+module Checkout
+  class FormDataAdapter
+    attr_reader :shipping_method_id
+
+    def initialize(params, order, current_user)
+      @params = params
+      @order = order
+      @current_user = current_user
+
+      move_payment_source_to_payment_attributes!
+
+      set_amount_in_payments_attributes
+
+      construct_saved_card_attributes if @params[:order][:existing_card_id]
+
+      @shipping_method_id = @params[:order].delete(:shipping_method_id)
+    end
+
+    def order_params
+      @params[:order]
+    end
+
+    private
+
+    # For payment step, filter order parameters to produce the expected
+    #   nested attributes for a single payment and its source,
+    #   discarding attributes for payment methods other than the one selected
+    def move_payment_source_to_payment_attributes!
+      return unless @params[:payment_source].present? &&
+                    payment_source_params = delete_payment_source_params!
+
+      @params[:order][:payments_attributes].first[:source_attributes] = payment_source_params
```

Comment:
It is strongly encouraged not to mutate the params object as can lead to potential and hard to spot bugs (I can try to find an article I read about the recently). This is just the same idea of not mutating data shared by multiple objects in a system. I'd rather `#dup` the params and return that one instead.

Reply:
ok, I am dupping the params on the initializer so that all work is done of a copy of it 👌

Reply:
this is now the only change in this PR :-) all other logic should be exactly the same.